File: src/core/combat/passives/compile.ts
// 룰 컴파일 — 보유 스킬의 passives + 캐릭터 traitIds → CompiledRule[] (전투 시작 시 makeUnit이 호출).
import type { CompiledRule, PassiveRule } from "../../types.ts";
import { SKILLS } from "../../../data/skills.ts";
import { CHARACTERS } from "../../../data/characters.ts";
import { TRAITS } from "../../../data/traits.ts";

function mk(rule: PassiveRule, via: { kind: "skill" | "trait" | "node"; id: string }, idx: number): CompiledRule {
  return { rule, via, idx, firedThisTurn: 0, firedThisBattle: 0 };
}

/** 스킬 패시브 = 활성(출전) 기준 — `skillIds`는 그 유닛의 활성 스킬. 특성(traitIds)은 항상 적용. */
export function compileRules(charId: string, skillIds: string[]): CompiledRule[] {
  const out: CompiledRule[] = [];
  let idx = 0;
  for (const sid of skillIds) {
    const sk = SKILLS[sid];
    if (!sk?.passives) continue;
    for (const rule of sk.passives) out.push(mk(rule, { kind: "skill", id: sid }, idx++));
  }
  for (const tid of CHARACTERS[charId]?.traitIds ?? []) {
    const t = TRAITS[tid];
    if (!t) continue;
    for (const rule of t.rules) out.push(mk(rule, { kind: "trait", id: tid }, idx++));
  }
  return out;
}

/** 노드 트리거 룰(인라인 PassiveRule[]) → CompiledRule[] (Phase C — createBattle가 한 유닛에 주입). */
export function compileInline(rules: PassiveRule[]): CompiledRule[] {
  return rules.map((rule, i) => mk(rule, { kind: "node", id: "node" }, i));
}

File: engine/spr-core/src/passives/compile.rs
//! 룰 컴파일 (TS `passives/compile.ts`) — 스킬 passives(활성) + 캐릭터 traitIds → CompiledRule[].
use spr_types::combat::CompiledRule;
use spr_types::data::Character;
use spr_types::passives::{PassiveRule, TraitDef};
use spr_types::skills::Skill;
use std::collections::HashMap;

fn mk(rule: PassiveRule, via_kind: &str, via_id: &str, idx: i64) -> CompiledRule {
    CompiledRule { rule, via_kind: via_kind.to_string(), via_id: via_id.to_string(), idx, fired_this_turn: 0, fired_this_battle: 0 }
}

/// 스킬 패시브(활성 기준) + 특성(항상) + 전직 부여 패시브(extra_trait_ids, 4.7). TS compileRules.
/// extra_trait_ids = 런 중 전직으로 부여된 trait id(`PartyMemberState.job_trait_ids`). 캐릭 traitIds 뒤에 이어 컴파일. 비전투/적은 &[].
pub fn compile_rules(
    char_id: &str,
    skill_ids: &[String],
    extra_trait_ids: &[String],
    chars: &HashMap<String, Character>,
    skills: &HashMap<String, Skill>,
    traits: &HashMap<String, TraitDef>,
) -> Vec<CompiledRule> {
    let mut out = Vec::new();
    let mut idx = 0i64;
    for sid in skill_ids {
        if let Some(sk) = skills.get(sid) {
            if let Some(ps) = &sk.passives {
                for rule in ps {
                    out.push(mk(rule.clone(), "skill", sid, idx));
                    idx += 1;
                }
            }
        }
    }
    if let Some(c) = chars.get(char_id) {
        for tid in &c.trait_ids {
            if let Some(t) = traits.get(tid) {
                for rule in &t.rules {
                    out.push(mk(rule.clone(), "trait", tid, idx));
                    idx += 1;
                }
            }
        }
    }
    // 전직 부여 패시브(런 한정) — 캐릭 특성 뒤에 결정론적 순서로.
    for tid in extra_trait_ids {
        if let Some(t) = traits.get(tid) {
            for rule in &t.rules {
                out.push(mk(rule.clone(), "trait", tid, idx));
                idx += 1;
            }
        }
    }
    out
}

/// 노드 트리거 룰(인라인) → CompiledRule[]. TS compileInline.
pub fn compile_inline(rules: &[PassiveRule]) -> Vec<CompiledRule> {
    rules.iter().enumerate().map(|(i, r)| mk(r.clone(), "node", "node", i as i64)).collect()
}
